### PR TITLE
Delayed bug audit

### PR DESCRIPTION
**Pull request by nnethercote**
I went through all the calls to `delayed_bug` and `span_delayed_bug` and found a few places where they could be avoided.

r? @compiler-errors

### CODE REVIEW

File: compiler/rustc_ast_lowering/src/delegation.rs [hunk 1]
```diff
@@ -103,12 +103,9 @@ impl<'hir> LoweringContext<'_, 'hir> {
...
         let sig_id = if self.is_in_trait_impl { item_id } else { path_id };
-        let sig_id = self
-            .resolver
-            .get_partial_res(sig_id)
-            .map(|r| r.expect_full_res().opt_def_id())
-            .unwrap_or(None);
-
+        let sig_id: Option<Option<DefId>> =
+            self.resolver.get_partial_res(sig_id).map(|r| r.expect_full_res().opt_def_id());
+        let sig_id: Option<DefId> = sig_id.unwrap_or(None);
```
Comment (kadiwa4) [inline compiler/rustc_ast_lowering/src/delegation.rs:108]:
```suggestion
        let sig_id: Option<DefId> =
            self.resolver.get_partial_res(sig_id).and_then(|r| r.expect_full_res().opt_def_id());
```

Comment (nnethercote) [reply compiler/rustc_ast_lowering/src/delegation.rs]: Done, thanks for the suggestion.

**Comment by nnethercote** [comment]
I rebased.

**Comment by compiler-errors** [comment]
Sorry, I'm kind of swamped and don't currently have the patience to review this as closely as it deserves

r? compiler

File: compiler/rustc_hir_typeck/src/fn_ctxt/_impl.rs [hunk 2]
```diff
@@ -857,12 +857,14 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                 Ok(r)
             })
             .or_else(|error| {
-                let guar = self
-                    .dcx()
-                    .span_delayed_bug(span, "method resolution should've emitted an error");
                 let result = match error {
                     method::MethodError::PrivateMatch(kind, def_id, _) => Ok((kind, def_id)),
-                    _ => Err(guar),
+                    _ => {
```
Comment (lcnr) [inline compiler/rustc_hir_typeck/src/fn_ctxt/_impl.rs:862]: please revert this change. I want to make sure we always fail if this returns `Err` and with your change this code could theoretically end up returning `Ok` for a private method and compilation may still succeed as we mess up reporting an error elsewhere.

File: compiler/rustc_infer/src/infer/error_reporting/mod.rs [hunk 3]
```diff
@@ -437,90 +439,96 @@ impl<'tcx> TypeErrCtxt<'_, 'tcx> {
...
-                    // attempt to do is go through a number of specific
-                    // scenarios and try to find the best way to present
-                    // the error. If all of these fails, we fall back to a rather
-                    // general bit of code that displays the error information
-                    RegionResolutionError::ConcreteFailure(origin, sub, sup) => {
-                        if sub.is_placeholder() || sup.is_placeholder() {
-                            self.report_placeholder_failure(origin, sub, sup).emit();
-                        } else {
-                            self.report_concrete_failure(origin, sub, sup).emit();
+        let guars: Vec<_> = errors
```
Comment (lcnr) [inline compiler/rustc_infer/src/infer/error_reporting/mod.rs:442]: instead of using a `Vec` :thinking: maybe just use an `Option` and overwrite it for each goal?

File: compiler/rustc_resolve/src/ident.rs [hunk 4]
```diff
@@ -1200,7 +1199,6 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
                                             }
                                         };
                                         self.report_error(span, error);
```
Comment (lcnr) [inline compiler/rustc_resolve/src/ident.rs:1201]:
```suggestion
                                        let _: ErrorGuaranteed = self.report_error(span, error);
```
it's non-trivial that `report_error` actually always reports an error so it would be good to propagate this outwards, iirc this code previously had a mode where it silently failed :thinking: apparently that has changed since then

File: compiler/rustc_hir_typeck/src/fn_ctxt/_impl.rs (hunk 2 above)
Comment (nnethercote) [reply compiler/rustc_hir_typeck/src/fn_ctxt/_impl.rs:862]: ok

File: compiler/rustc_infer/src/infer/error_reporting/mod.rs (hunk 3 above)
Comment (nnethercote) [reply compiler/rustc_infer/src/infer/error_reporting/mod.rs:442]: ok

File: compiler/rustc_middle/src/ty/adt.rs [hunk 5]
```diff
@@ -494,11 +494,15 @@ impl<'tcx> AdtDef<'tcx> {
...
-                    ErrorHandled::Reported(..) => "enum discriminant evaluation failed",
-                    ErrorHandled::TooGeneric(..) => "enum discriminant depends on generics",
+                match err {
+                    ErrorHandled::Reported(..) => {} // error guaranteed by `Reported`
+                    ErrorHandled::TooGeneric(..) => {
+                        tcx.dcx().span_delayed_bug(
+                            tcx.def_span(expr_did),
+                            "enum discriminant depends on generics",
+                        );
+                    }
```
[the remark this answers is not in the document]
Comment (nnethercote) [reply compiler/rustc_middle/src/ty/adt.rs:504]: ok

File: compiler/rustc_resolve/src/ident.rs (hunk 4 above)
Comment (nnethercote) [reply compiler/rustc_resolve/src/ident.rs:1201]: I've made the suggested change. It would be nice if `Res::Err` could contain an `ErrorGuaranteed`, but there are many places where one is constructed without a guarantee being available.

**Comment by nnethercote** [comment]
New code is up, hopefully it addresses all the comments.

**Comment by nnethercote** [comment]
This wasn't expected to have any effect on performance. I suspect the (good) results are just noise.
